vllm/v1/core/guided_decoding/__init__.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import annotations
+
+import copy
+import threading
+from abc import ABC, abstractmethod
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Generic, Literal, TypeVar, get_args
+
+from transformers import PreTrainedTokenizer
+
+from vllm.config import ModelConfig
+from vllm.logger import init_logger
+from vllm.utils import LazyLoader
+from vllm.v1.request import GuidedDecodingKey, Request, RequestStatus
+
+from .grammar import Grammar
+
+if TYPE_CHECKING:
+    import xgrammar as xgr
+    from transformers import PreTrainedTokenizer
+    from typing_extensions import LiteralString
+
+    from vllm.transformers_utils.tokenizer_group import BaseTokenizerGroup
+
+    from .grammar import XGrammar
+else:
+    xgr = LazyLoader("xgr", globals(), "xgrammar")
+
+logger = init_logger(__name__)
+
+__all__ = ["Grammar", "GuidedDecodingManager"]
+
+
+@dataclass
+class GrammarCache:
+    value: Grammar | None
+    event: threading.Event
+
+
+T = TypeVar("T", bound=str)
+
+
+class GuidedDecodingManager(ABC, Generic[T]):
+
+    @abstractmethod
+    def initialize_cache(self, key: GuidedDecodingKey) -> Grammar:
+        ...
+
+    def flush(self):
+        with self._lock:
+            self.grammar_cache.clear()
+
+    def cache(self, request: Request):
```

Comment:
I'm curious what you hit that made you need your own custom thread-safe cache implementation? Did you consider using functools and hit a reason it wasn't flexible enough?  https://docs.python.org/3/library/functools.html

Reply:
On the scheduler, we need to know when the request grammar is ready to then appendleft to the queue.

I was initially using lru_cache, but end up discarding it given that we have to track the compilation finish some how.